streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -368,50 +378,132 @@ void handleLostAll() {
-                } catch (final TaskIdFormatException e) {
-                    // there may be some unknown files that sits in the same directory,
-                    // we should ignore these files instead trying to delete them as well
+                } catch (final IOException e) {
+                    // if for any reason we can't lock this task dir, just move on
+                    log.warn(String.format("Exception caught while attempting to lock task %s:", id), e);
+                }
+            } catch (final TaskIdFormatException e) {
+                // ignore any unknown files that sit in the same directory
+            }
+        }
+    }
+
+    /**
+     * Attempts to release the lock for the passed in task's directory. It does not remove the task from
+     * {@code lockedTaskDirectories} so it's safe to call during iteration, and should be idempotent.
+     */
+    private RuntimeException releaseTaskDirLock(final TaskId taskId) {
+        try {
+            stateDirectory.unlock(taskId);
+        } catch (final IOException e) {
+            log.warn(String.format("Caught the following exception while trying to unlock task %s", taskId), e);
+            return new StreamsException(String.format("Unable to unlock task directory %s", taskId), e);
+        }
+        return null;
+    }
+
+    /**
+     * We must release the lock for any unassigned tasks that we temporarily locked in preparation for a
+     * rebalance in {@link #tryToLockAllTaskDirectories()}.
+     */
+    private void releaseLockedUnassignedTaskDirectories() {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+
+        final Iterator<TaskId> taskIdIterator = lockedTaskDirectories.iterator();
+        while (taskIdIterator.hasNext()) {
+            final TaskId id = taskIdIterator.next();
+
+            if (!tasks.containsKey(id)) {
+                final RuntimeException unlockException = releaseTaskDirLock(id);
+                if (unlockException == null) {
+                    taskIdIterator.remove();
+                } else {
+                    log.error("Failed to release the lock for task directory {}.", id);
+                    firstException.compareAndSet(null, unlockException);
+                }
+            }
+        }
+
+        final RuntimeException fatalException = firstException.get();
+        if (fatalException != null) {
+            throw fatalException;
+        }
+    }
+
+    private long sumOfChangelogOffsets(final Map<TopicPartition, Long> changelogOffsets) {
+        long offsetSum = 0L;
+        for (final Map.Entry<TopicPartition, Long> changelogEntry : changelogOffsets.entrySet()) {
+            final TopicPartition changelog = changelogEntry.getKey();
+            final long offset = changelogEntry.getValue();
+
+            if (offset < 0L) {
+                if (offset == -1L) {
+                    log.debug("Skipping unknown offset for changelog {}", changelog);
```

Comment:
In what cases might one store's offsets being invalid mean that every other store with valid offsets should not be taken into account?

That's not rhetorical, I really am asking. It's not clear to me exactly when you'd get this invalid offset response -- but if only one partition was having issues (whatever those may be) and the others al had valid, positive offsets, would we have to wipe out the entire state? Do we even check for negative offsets elsewhere in Streams? It's not clear to me that/if we do (in fact several places assume they are always positive and I believe would actually crash if not)

Reply:
As far as I can see from the code, an offset of -1 is returned in error case. Additionally, offset -1 is used during initialization of the producer response. I guess in the error case we do not write any offset into the checkpoint file and the offset map. So I suppose that it cannot happen that the offset becomes -1 in this code. So, my proposal would be to double-check my observation. In case it is correct we can specify an invariant that the offset must be >= 0 and throw an `IllegalStateException` if the invariant is not satisfied.

Reply:
We skip adding offsets to the map in the error case, but I asked Jason and apparently there may be a non-error case where -1 is returned with an idempotent producer. Just cc'ed you on the thread